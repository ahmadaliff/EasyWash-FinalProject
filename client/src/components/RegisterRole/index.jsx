import PropTypes from 'prop-types';
import { useState } from 'react';
import { useForm } from 'react-hook-form';
import { connect, useDispatch } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { FormattedMessage, injectIntl } from 'react-intl';

import InputRHF from '@components/InputRHF';
import MapLeaflet from '@components/MapLeaflet';

import { styled } from '@mui/material/styles';
import { AccountBox, LocalLaundryService } from '@mui/icons-material';
import { FormControl, FormGroup, ToggleButton, ToggleButtonGroup } from '@mui/material';

import { selectIsVerify } from '@containers/Client/selectors';
import { actionSetMerchant, actionSetRole, actionSetStep } from '@pages/Register/actions';
import { selectEmail, selectMerchant, selectRole, selectStep } from '@pages/Register/selectors';

import classes from '@components/RegisterRole/style.module.scss';

const RegisterRole = ({ rolePersist, merchant, isVerified, step, intl: { formatMessage } }) => {
  const dispatch = useDispatch();
  const [role, setRole] = useState(rolePersist || 'user');
  const [location, setLocation] = useState();

  const handleChange = (e, value) => {
    setRole(value);
  };

  const {
    handleSubmit,
    register,
    formState: { errors },
    watch,
  } = useForm();
  const name = watch('name');
  const description = watch('description');

  const StyledToggleButton = styled(ToggleButton)({
    '&.Mui-selected, &.Mui-selected:hover': {
      color: 'white',
      backgroundColor: '#7ac94c',
      border: 'none',
    },
  });

  const handleLocation = (loc) => {
    setLocation(loc);
  };

  const onSubmit = (data) => {
    dispatch(actionSetRole(role));
    if (role !== 'user') {
      data.location = location && JSON.stringify(location);
      dispatch(actionSetMerchant(data));
    }
    dispatch(actionSetStep(3));
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)} data-testid="register-role">
      <FormGroup>
        <p className={classes.labelRole}>
          <FormattedMessage id="app_user_role" />
        </p>
        <ToggleButtonGroup
          value={role}
          exclusive
          onChange={handleChange}
          fullWidth
          className={classes.toggleRole}
          size="small"
          data-testid="role"
        >
          <StyledToggleButton data-testid="button-role-user" value="user">
            <AccountBox /> <FormattedMessage id="app_user" />
          </StyledToggleButton>
          <StyledToggleButton data-testid="button-role-merchant" value="merchant">
            <LocalLaundryService /> Laundry
          </StyledToggleButton>
        </ToggleButtonGroup>
      </FormGroup>
      {role !== 'user' && (
        <>
          <h5 className={classes.merchantFormHeader}>Merchant Info</h5>
          <InputRHF
            input={{
              name: 'name',
              required: formatMessage({ id: 'app_merchant_require' }),
              type: 'text',
              label: formatMessage({ id: 'app_merchant_name' }),
              value: merchant?.name,
            }}
            register={register}
            errors={errors}
          />
          <FormControl fullWidth>
            <div className={classes.inputLabel}>
              <FormattedMessage id="app_location" />
              <MapLeaflet handleLocation={handleLocation} islocated={merchant?.location} permanent={false} />
            </div>
          </FormControl>
          <FormControl className={classes.inputContainer}>
            <label htmlFor="description" className={classes.inputLabel}>
              <FormattedMessage id="app_description" />
              <textarea
                className={`${classes.inputTextArea} ${errors.description && classes.inputError}`}
                {...register('description', {})}
                defaultValue={merchant?.description}
                data-testid="input-description"
              />
            </label>
          </FormControl>
        </>
      )}
      <div className={classes.buttonWrap}>
        <button
          type="button"
          className={classes.buttonSubmit}
          onClick={() => (isVerified ? dispatch(actionSetStep(step - 2)) : dispatch(actionSetStep(step - 1)))}
        >
          <FormattedMessage id="app_back" />
        </button>
        <button
          type="submit"
          className={classes.buttonSubmit}
          disabled={role !== 'user' ? !role || !name || !description : !role}
          data-testid="button-submit"
        >
          <FormattedMessage id="app_next" />
        </button>
      </div>
    </form>
  );
};

RegisterRole.propTypes = {
  intl: PropTypes.object,
  step: PropTypes.number,
  email: PropTypes.string,
  rolePersist: PropTypes.string,
  isVerified: PropTypes.bool,
  merchant: PropTypes.object,
};

const mapStateToProps = createStructuredSelector({
  step: selectStep,
  email: selectEmail,
  isVerified: selectIsVerify,
  rolePersist: selectRole,
  merchant: selectMerchant,
});

export default injectIntl(connect(mapStateToProps)(RegisterRole));
