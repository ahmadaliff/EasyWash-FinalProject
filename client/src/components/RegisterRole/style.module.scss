@import '@styles/base';

.showPassword,
.nav {
  @include directionX;
  align-items: center;
  padding-top: 0.25rem;
  font-size: small;
  color: var(--color-primary);
}
.nav {
  color: var(--color-text);
  a {
    color: var(--color-primary);
  }
}
.toggleRole {
  button {
    color: var(--color-primary);
    text-transform: capitalize;
    border-color: var(--color-primary);
  }
}
.labelRole {
  margin: 0;
  padding-bottom: 0.25rem;
  font-size: small;
}
.merchantFormHeader {
  margin: 1rem 0 0.25rem 0;
}
.buttonWrap {
  @include directionX;
  justify-content: space-between;
  gap: 0.5rem;
  padding-top: 1rem;
  .buttonSubmit {
    width: 100%;
    background-color: var(--color-primary);
    color: var(--color-bg);
    border-radius: 2.5rem;
    padding: 0.5rem 2.5rem 0.5rem 2.5rem;
    margin: auto;
    margin-top: 0.75rem;
    box-shadow: none;
    border: none;
    cursor: pointer;
    &:hover {
      transform: scale(1.0125, 1.0125);
    }
    &:disabled {
      background-color: rgba(gray, 0.2);
      color: rgba($color: gray, $alpha: 0.5);
    }
    &:disabled:hover {
      transform: none;
    }
  }
}
.inputContainer {
  width: 100%;
  .inputTextArea {
    box-sizing: border-box;
    width: 100%;
    min-height: 5rem;
    padding: 0.5rem 1rem 0.5rem 1rem;
    font-size: medium;
    border: 1px solid var(--color-text);
    background-color: var(--color-bg);
    color: var(--color-text);
    border-radius: 10px;
    &:focus {
      outline: none;
      border: 1px solid var(--color-primary);
    }
  }
  .inputError {
    border: 1px solid #ed0226;
  }
}

.inputLabel {
  @include directionY;
  gap: 0.25rem;
  width: 100%;
  padding: 0.5rem 0 0.5rem 0;
  font-weight: 400;
  font-size: small;
}
.inputLabelError {
  color: #ed0226;
  font-size: small;
}
