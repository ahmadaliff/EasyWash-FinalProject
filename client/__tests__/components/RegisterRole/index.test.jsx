import RegisterRole from '@components/RegisterRole';
import { fireEvent, render, screen } from '@utils/testHelper';
import { act } from 'react-dom/test-utils';

let wrapper;
let inputName;
let inputDescription;

const mockProps = {
  intl: { formatMessage: jest.fn() },
  step: 10,
};

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  useNavigate: () => mockNavigate,
}));
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => jest.fn(),
}));
jest.mock('@pages/Register/selectors', () => ({
  ...jest.requireActual('@pages/Register/selectors'),
  selectRole: () => 'merchant',
}));

jest.mock('@components/MapLeaflet', () => () => <div>MockMapLeaflet</div>);

beforeEach(() => {
  act(() => {
    wrapper = render(<RegisterRole {...mockProps} />);
  });
});
describe('RegisterRole Component', () => {
  test('Correct render', () => {
    const { getByTestId } = wrapper;
    expect(getByTestId('register-role')).toBeInTheDocument();
  });
  test('Should change role state', () => {
    const { getByTestId } = wrapper;
    const button = getByTestId('button-role-merchant');
    fireEvent.click(button);
  });
  test('Should submit properly with role merchant', async () => {
    const { getByTestId } = wrapper;
    inputName = screen.getByLabelText('Nama Laundry');
    inputDescription = getByTestId('input-description');
    fireEvent.change(inputName, { target: { value: 'Laundry Mantap' } });
    fireEvent.change(inputDescription, { target: { value: 'ini adlaah deskripsi' } });
    expect(inputName.value).toBe('Laundry Mantap');
    expect(inputDescription.value).toBe('ini adlaah deskripsi');
    const button = getByTestId('button-submit');
    await act(() => {
      fireEvent.click(button);
    });
  });
  test('Should match with snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
