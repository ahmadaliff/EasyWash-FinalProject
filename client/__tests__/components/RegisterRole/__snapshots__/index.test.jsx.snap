// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterRole Component Should match with snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        data-testid="register-role"
      >
        <div
          class="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
        >
          <p
            class="labelRole"
          >
            Daftar sebagai 
          </p>
          <div
            class="MuiToggleButtonGroup-root MuiToggleButtonGroup-fullWidth toggleRole css-mgs850-MuiToggleButtonGroup-root"
            data-testid="role"
            role="group"
          >
            <button
              aria-pressed="false"
              class="MuiButtonBase-root MuiToggleButton-root MuiToggleButton-fullWidth MuiToggleButton-sizeSmall MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-x7mo1m-MuiButtonBase-root-MuiToggleButton-root"
              data-testid="button-role-user"
              tabindex="0"
              type="button"
              value="user"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="AccountBoxIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm-7 3c1.93 0 3.5 1.57 3.5 3.5S13.93 13 12 13s-3.5-1.57-3.5-3.5S10.07 6 12 6zm7 13H5v-.23c0-.62.28-1.2.76-1.58C7.47 15.82 9.64 15 12 15s4.53.82 6.24 2.19c.48.38.76.97.76 1.58V19z"
                />
              </svg>
               
              Pengguna
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              aria-pressed="true"
              class="MuiButtonBase-root MuiToggleButton-root Mui-selected MuiToggleButton-fullWidth MuiToggleButton-sizeSmall MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-x7mo1m-MuiButtonBase-root-MuiToggleButton-root"
              data-testid="button-role-merchant"
              tabindex="0"
              type="button"
              value="merchant"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="LocalLaundryServiceIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M9.17 16.83c1.56 1.56 4.1 1.56 5.66 0 1.56-1.56 1.56-4.1 0-5.66l-5.66 5.66zM18 2.01 6 2c-1.11 0-2 .89-2 2v16c0 1.11.89 2 2 2h12c1.11 0 2-.89 2-2V4c0-1.11-.89-1.99-2-1.99zM10 4c.55 0 1 .45 1 1s-.45 1-1 1-1-.45-1-1 .45-1 1-1zM7 4c.55 0 1 .45 1 1s-.45 1-1 1-1-.45-1-1 .45-1 1-1zm5 16c-3.31 0-6-2.69-6-6s2.69-6 6-6 6 2.69 6 6-2.69 6-6 6z"
                />
              </svg>
               Laundry
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        <h5
          class="merchantFormHeader"
        >
          Merchant Info
        </h5>
        <div
          class="MuiFormControl-root inputContainer css-1nrlq1o-MuiFormControl-root"
          data-testid="inputRHF"
        >
          <label
            class="inputLabel"
            for="name"
          >
            Nama Laundry
            <input
              class="input undefined"
              id="name"
              name="name"
              type="text"
              value=""
            />
          </label>
        </div>
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
        >
          <div
            class="inputLabel"
          >
            Lokasi
            <div>
              MockMapLeaflet
            </div>
          </div>
        </div>
        <div
          class="MuiFormControl-root inputContainer css-1nrlq1o-MuiFormControl-root"
        >
          <label
            class="inputLabel"
            for="description"
          >
            Deskripsi
            <textarea
              class="inputTextArea undefined"
              data-testid="input-description"
              name="description"
            />
          </label>
        </div>
        <div
          class="buttonWrap"
        >
          <button
            class="buttonSubmit"
            type="button"
          >
            Kembali
          </button>
          <button
            class="buttonSubmit"
            data-testid="button-submit"
            disabled=""
            type="submit"
          >
            Lanjut
          </button>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      data-testid="register-role"
    >
      <div
        class="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
      >
        <p
          class="labelRole"
        >
          Daftar sebagai 
        </p>
        <div
          class="MuiToggleButtonGroup-root MuiToggleButtonGroup-fullWidth toggleRole css-mgs850-MuiToggleButtonGroup-root"
          data-testid="role"
          role="group"
        >
          <button
            aria-pressed="false"
            class="MuiButtonBase-root MuiToggleButton-root MuiToggleButton-fullWidth MuiToggleButton-sizeSmall MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-x7mo1m-MuiButtonBase-root-MuiToggleButton-root"
            data-testid="button-role-user"
            tabindex="0"
            type="button"
            value="user"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="AccountBoxIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm-7 3c1.93 0 3.5 1.57 3.5 3.5S13.93 13 12 13s-3.5-1.57-3.5-3.5S10.07 6 12 6zm7 13H5v-.23c0-.62.28-1.2.76-1.58C7.47 15.82 9.64 15 12 15s4.53.82 6.24 2.19c.48.38.76.97.76 1.58V19z"
              />
            </svg>
             
            Pengguna
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            aria-pressed="true"
            class="MuiButtonBase-root MuiToggleButton-root Mui-selected MuiToggleButton-fullWidth MuiToggleButton-sizeSmall MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-x7mo1m-MuiButtonBase-root-MuiToggleButton-root"
            data-testid="button-role-merchant"
            tabindex="0"
            type="button"
            value="merchant"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="LocalLaundryServiceIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M9.17 16.83c1.56 1.56 4.1 1.56 5.66 0 1.56-1.56 1.56-4.1 0-5.66l-5.66 5.66zM18 2.01 6 2c-1.11 0-2 .89-2 2v16c0 1.11.89 2 2 2h12c1.11 0 2-.89 2-2V4c0-1.11-.89-1.99-2-1.99zM10 4c.55 0 1 .45 1 1s-.45 1-1 1-1-.45-1-1 .45-1 1-1zM7 4c.55 0 1 .45 1 1s-.45 1-1 1-1-.45-1-1 .45-1 1-1zm5 16c-3.31 0-6-2.69-6-6s2.69-6 6-6 6 2.69 6 6-2.69 6-6 6z"
              />
            </svg>
             Laundry
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <h5
        class="merchantFormHeader"
      >
        Merchant Info
      </h5>
      <div
        class="MuiFormControl-root inputContainer css-1nrlq1o-MuiFormControl-root"
        data-testid="inputRHF"
      >
        <label
          class="inputLabel"
          for="name"
        >
          Nama Laundry
          <input
            class="input undefined"
            id="name"
            name="name"
            type="text"
            value=""
          />
        </label>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
      >
        <div
          class="inputLabel"
        >
          Lokasi
          <div>
            MockMapLeaflet
          </div>
        </div>
      </div>
      <div
        class="MuiFormControl-root inputContainer css-1nrlq1o-MuiFormControl-root"
      >
        <label
          class="inputLabel"
          for="description"
        >
          Deskripsi
          <textarea
            class="inputTextArea undefined"
            data-testid="input-description"
            name="description"
          />
        </label>
      </div>
      <div
        class="buttonWrap"
      >
        <button
          class="buttonSubmit"
          type="button"
        >
          Kembali
        </button>
        <button
          class="buttonSubmit"
          data-testid="button-submit"
          disabled=""
          type="submit"
        >
          Lanjut
        </button>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
